sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala

```diff
@@ -637,4 +637,40 @@ object IntervalUtils {
 
     new CalendarInterval(totalMonths, totalDays, micros)
   }
+
+  /**
+   * Adjust interval so 30-day time periods are represented as months
+   */
+  def justifyDays(interval: CalendarInterval): CalendarInterval = {
+    val monthToDays = interval.months * DAYS_PER_MONTH
+    val totalDays = Math.addExact(monthToDays, interval.days)
+    val months = Math.toIntExact(totalDays / DAYS_PER_MONTH)
+    val days = Math.toIntExact(totalDays % DAYS_PER_MONTH)
+    new CalendarInterval(months, days, interval.microseconds)
+  }
+
+  /**
+   * Adjust interval so 24-hour time periods are represented as days
+   */
+  def justifyHours(interval: CalendarInterval): CalendarInterval = {
+    val dayToUs = MICROS_PER_DAY * interval.days
+    val totalUs = Math.addExact(interval.microseconds, dayToUs)
+    val days = Math.toIntExact(totalUs / MICROS_PER_DAY)
+    val microseconds = totalUs % MICROS_PER_DAY
+    new CalendarInterval(interval.months, days, microseconds)
+  }
+
+  /**
+   * Adjust interval using justifyHours and justifyDays, with additional sign adjustments
+   */
+  def justifyInterval(interval: CalendarInterval): CalendarInterval = {
+    val monthToDays = DAYS_PER_MONTH * interval.months
+    val dayToUs = Math.multiplyExact(monthToDays + interval.days, MICROS_PER_DAY)
+    val totalUs = Math.addExact(interval.microseconds, dayToUs)
+    val microseconds = totalUs % MICROS_PER_DAY
+    val totalDays = totalUs / MICROS_PER_DAY
+    val days = totalDays % DAYS_PER_MONTH
+    val months = Math.toIntExact(totalDays / DAYS_PER_MONTH)
```

Comment:
The worse case is `Long.Max / MICROS_PER_DAY  / DAYS_PER_MONTH`, which can't overflow.

Reply:
ok, i'd remove these useless overflow check